Add design docs (#2657)
open-telemetry/opentelemetry-rust: docs/design/logs.md

# OpenTelemetry Rust Logs Design

Status:

> [Development](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)

Author: Both.

> [Development](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/document-status.md)

Reviewer: I reckon it'd be good to link to the [otel spec logs doc](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/logs) as well

## Overview

OpenTelemetry (OTel) Logs support differs from Metrics and Traces as it does not
introduce a new logging API for end users. Instead, OTel recommends leveraging

> existing logging libraries such as `log` and `tracing`, while providing bridges

Reviewer: ```suggestion
existing logging libraries such as [log](https://crates.io/crates/log) or `[tracing](https://crates.io/crates/tracing)`, while providing bridges
```

Reviewer: It might also be worth pointing out that tracing is not "distributed tracing", but rather in-process-structured-logging-and-execution-traces.

Because the name makes it easy to think it is something it is not

> (appenders) to route logs through OpenTelemetry.

Author: Not really as spec removed the "only for bridge authors" part and rename LogBridge API to Log API.

> Unlike Traces and Metrics, which introduced new APIs, Logs took a different

Reviewer: ```suggestion
```

Reviewer: I think you can consolidate the first two paragraphs here

> approach due to the long history of existing logging solutions. In Rust, the

Reviewer: ```suggestion
In Rust, these
```

> most widely used logging libraries are `log` and `tracing`. OTel Rust maintains

Reviewer: ```suggestion
are the most widely used logging libraries, and have been embraced in the community for some time. OTel Rust maintains
```

appenders for these libraries, allowing users to seamlessly integrate with
OpenTelemetry without changing their existing logging instrumentation.

The `tracing` appender is particularly optimized for performance due to its
widespread adoption and the fact that `tracing` itself has a bridge from the
`log` crate. Notably, OpenTelemetry Rust itself is instrumented using `tracing`
for internal logs. Additionally, when OTel began supporting logging as a signal,
the `log` crate lacked structured logging support, reinforcing the decision to
prioritize `tracing`.

## Benefits of OpenTelemetry Logs

- **Unified configuration** across Traces, Metrics, and Logs.

> - **Automatic correlation** with Traces.

Author: https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/logs/logger.rs#L41-L50 It is not under feature flag?

- **Consistent Resource attributes** across signals.
- **Multiple destinations support**: Logs can continue flowing to existing

>   destinations like stdout while also being sent to an OpenTelemetry-capable

Author: I meant stdout (and other destinations) *outside* of OTel. eg: Tracing::fmt layer. or journal layer etc.

  backend, typically via an OTLP Exporter or exporters that export to operating
  system native systems like `Windows ETW` or `Linux user_events`.
- **Standalone logging support** for applications that use OpenTelemetry as
  their primary logging mechanism.

## Key Design Principles

- High performance - no locks/contention in the hot path, minimal/no heap
  allocation.
- Capped resource usage - well-defined behavior when overloaded.

> - Self-observable.

Reviewer: Could you clarify this (I assume "the logs subsystem logs about its own work", but not 100%)

> - Well defined Error handling, returning Result as appropriate instead of panic.

Reviewer: ```suggestion
- Robust Error handling, returning Result where possible instead of panicing
```

"Robust" I think works well - it makes it clear what the goal is

- Minimal public API, exposing based on need only.

> 

Reviewer: We could chuck a diagram in to help the reader put the pieces together in their mind - you can [embed mermaidjs directly in Github markdown files](https://github.blog/developer-skills/github/include-diagrams-markdown-files-mermaid/). Here's an extremely rough sketch of what that could look like 

```suggestion

## Architecture Overview
```mermaidjs
graph TD
    subgraph Application
        A1[Application Code]
    end

    subgraph Logging Libraries
        B1[log Crate]
        B2[tracing Crate]
    end

    subgraph OpenTelemetry
        C1[OpenTelemetry Appender for log]
        C2[OpenTelemetry Appender for tracing]
        C3[OpenTelemetry Logs API]
        C4[OpenTelemetry Logs SDK]
        C5[OTLP Exporter]
    end

    subgraph Observability Backend
        D1[OTLP-Compatible Backend]
    end

    A1 --> |Emits Logs| B1
    A1 --> |Emits Logs| B2
    B1 --> |Bridged by| C1
    B2 --> |Bridged by| C2
    C1 --> |Sends to| C3
    C2 --> |Sends to| C3
    C3 --> |Processes with| C4
    C4 --> |Exports via| C5
    C5 --> |Sends to| D1
```

## Logs API

The OTel Logs API is not intended for direct end-user usage. Instead, it is
designed for appender/bridge authors to integrate existing logging libraries
with OpenTelemetry. However, there is nothing preventing it from being used by
end-users.

### API Components

> 

Reviewer: Would be helpful to have links into the repo here - for discoverability, and also because sometimes its not obvious where stuff is (e.g. it took me a while to realise `./opentelemetry` contains the API!)

1. **Key-Value Structs**: Used in `LogRecord`, where keys are shared across
   signals but values differ from Metrics and Traces. This is because values in
   Logs can contain more complex structures than those in Traces and Metrics.
2. **Traits**:

>     - `LoggerProvider` - provides methods to obtain Logger.

Reviewer: Things are getting stable enough with the RC that we could probably also risk linking out to these - e.g.

```suggestion
    - [LoggerProvider](https://github.com/open-telemetry/opentelemetry-rust/blob/05920759303358cb979b7b5538da2b6b5c0be8ee/opentelemetry/src/logs/logger.rs#L28)  - provides methods to obtain a `Logger`.
```

    - `Logger` - provides methods to create LogRecord and emit the created
      LogRecord.
    - `LogRecord` - provides methods to populate LogRecord.
3. **No-Op Implementations**: By default, the API performs no operations until
   an SDK is attached.

### Logs Flow

1. Obtain a `LoggerProvider` implementation.
2. Use the `LoggerProvider` to create `Logger` instances, specifying a scope
   name (module/component emitting logs). Optional attributes and version are
   also supported.
3. Use the `Logger` to create an empty `LogRecord` instance.
4. Populate the `LogRecord` with body, timestamp, attributes, etc.
5. Call `Logger.emit(LogRecord)` to process and export the log.

If only the Logs API is used (without an SDK), all the above steps result in no
operations, following OpenTelemetry’s philosophy of separating API from SDK. The
official Logs SDK provides real implementations to process and export logs.
Users or vendors can also provide alternative SDK implementations.

## Logs SDK

The OpenTelemetry Logs SDK provides an OTel specification-compliant
implementation of the Logs API, handling log processing and export.

### Core Components

#### `SdkLoggerProvider`

> 

Reviewer: IMHO it'd be helpful to have a 1-liner purpose for each of these, e.g.:

```suggestion
Allows bridge authors to obtain `Logger` instances. 
```

or so

- Implements the `LoggerProvider` trait.
- Creates and manages `SdkLogger` instances.
- Holds logging configuration, including `Resource` and processors.
- Does not retain a list of created loggers. Instead, it passes an owned clone
  of itself to each logger created. This is done so that loggers get a hold of
  the configuration (like which processor to invoke).
- Uses an `Arc<LoggerProviderInner>` and delegates all configuration to
  `LoggerProviderInner`. This allows cheap cloning of itself and ensures all
  clones point to the same underlying configuration.
- As `SdkLoggerProvider` only holds an `Arc` of its inner, it can only accept

>   `&self` in its methods like flush and shutdown. Else it needs to rely on

Reviewer 2: nit - just to add more clarity `Since SdkLoggerProvider holds an Arc to its inner state, it can take &self in methods like flush and shutdown, enabling shared access without requiring exclusive ownership. Without Arc, these methods would need interior mutability, which comes with runtime performance costs`

  interior mutability that comes with runtime performance costs. Since methods
  like shutdown usually need to mutate interior state, components like exporter
  use interior mutability to handle shutdown. (More on this in the exporter
  section)
- An alternative design was to let `SdkLogger` hold a `Weak` reference to the
  `SdkLoggerProvider`. This would be a `weak->arc` upgrade in every log emission,
  significantly affecting throughput.

> - `LoggerProviderInner` implements `Drop`, triggering `shutdown()` when no

Reviewer 3: Is there a way, for single threaded async runtimes, to introduce `shutdown_async`  method that does not block? After this call, the shutdown flag is set and we can be sure no deadlock can happen. 

The same thing is applicable for metrics nad traces.

Author: https://github.com/open-telemetry/opentelemetry-rust/issues/2476 Yes, we need to introduce a non-blocking version of shutdown in the future.

  references remain. However, in practice, loggers are often stored statically
  inside appenders (like tracing-appender), so explicit shutdown by the user is

>   required.

Reviewer 2: I think explicit shutdown is not always required. Perhaps we can add `That said, shutdown() will still be triggered automatically when the application exits, as tracing-appender is destroyed, or if the user explicitly drops the tracing-appender instance.`

Reviewer 2: Yes, this is true and quite unusual, because even global statics are dropped after main during application shutdown. This should theoretically ensure that all destructors/drop implementations are invoked.

Author: This is same if we use set_global_provider. Drop is not invoked there automatically as the instance passed to global is there forever...

#### `SdkLogger`

- Implements the `Logger` trait.
- Creates `SdkLogRecord` instances and emits them.
- Calls `OnEmit()` on all registered processors when emitting logs.
- Passes mutable references to each processor (`&mut log_record`), i.e.,
  ownership is not passed to the processor. This ensures that the logger avoids

>   cloning costs. Since a mutable reference is passed, processors can modify the

Reviewer 3: ```suggestion
...processors can modify...
```

Is this how enrichment can be implemented?

Additionally, is there a way to pass `Context` object when emitting the log that can be used by enrichers to add additional dimensions. 

The context object can be quite heavy, for example `HttpRequest`

Author: > Is this how enrichment can be implemented?

Yes this is one way to enrich Logs without having to modify the instrumentation side. Not just enrichment, but redaction/others too can be done here.

Author: Yes you can access the current context in the processor and use that to enrich. OTel SDK already uses Context to attach trace-id and span-id to the log.

  system native facilities like `Windows ETW` or `Linux user_events`.
- OTLP Exporting requires network calls (HTTP/gRPC) and batching of logs for
  efficiency purposes. These exporters buffer log records by copying them to the
  heap. (More on this in the BatchLogRecordProcessor section)

#### `LogRecord`

- Holds log data, including attributes.
- Uses an inline array for up to 5 attributes to optimize stack usage.
- Falls back to a heap-allocated `Vec` if more attributes are required.
- Inspired by Go’s `slog` library for efficiency.

#### LogRecord Processors

`SdkLoggerProvider` allows being configured with any number of LogProcessors.
They get called in the order of registration. Log records are passed to the
`OnEmit` method of LogProcessor. LogProcessors can be used to process the log
records, enrich them, filter them, and export to destinations by leveraging
LogRecord Exporters.

Following built-in Log processors are provided in the Log SDK:

##### SimpleLogProcessor

This processor is designed to be used for exporting purposes. Export is handled
by an Exporter (which is a separate component). SimpleLogProcessor is "simple"

> in the sense that it does not attempt to do any processing - it just calls the

Reviewer 3: ```suggestion
in the sense that it does not attempt to do any processing - it just calls the
```

this is ideal for ETW exporter, or not?

Author: We haven't closed how to model ETW/user-events yet. It could be modeled as an Exporter, and pair it with a Rentrant or SimpleConcurrentProcessor. Or they can be modeled as processor itself.

  like shutdown usually need to mutate interior state, but this component can
  only take `&self`, it defers to components like exporter to use interior
  mutability to handle shutdown. (More on this in the exporter section)
- An alternative design was to let `SdkLogger` hold a `Weak` reference to the
  `SdkLoggerProvider`. This would be a `weak->arc` upgrade in every log
  emission, significantly affecting throughput.
- `LoggerProviderInner` implements `Drop`, triggering `shutdown()` when no
  references remain. However, in practice, loggers are often stored statically
  inside appenders (like tracing-appender), so explicit shutdown by the user is
  required.

#### `SdkLogger`

This is an implementation of the `Logger`, and contains functionality to create
and emit logs.

- Implements the `Logger` trait.
- Creates `SdkLogRecord` instances and emits them.
- Calls `OnEmit()` on all registered processors when emitting logs.
- Passes mutable references to each processor (`&mut log_record`), i.e.,
  ownership is not passed to the processor. This ensures that the logger avoids
  cloning costs. Since a mutable reference is passed, processors can modify the

>   log, and it will be visible to the next processor in the chain.

Reviewer 2: Nothing to update here, but we need #2103 to get merged to have this feature fully functional :)

Author: Yes that reminds me to get that change merged soon!

- Since the processor only gets a reference to the log, it cannot store it
  beyond the `OnEmit()`. If a processor needs to buffer logs, it must explicitly
  copy them to the heap.
- This design allows for stack-only log processing when exporting to operating
  system native facilities like `Windows ETW` or `Linux user_events`.
- OTLP Exporting requires network calls (HTTP/gRPC) and batching of logs for
  efficiency purposes. These exporters buffer log records by copying them to the

>   heap. (More on this in the BatchLogRecordProcessor section)

Reviewer 2: Also good to mention about the `logger::enabled` function, which enables logger/processor/exporter to decide whether pipeline is enabled for given log metadata, and so avoid costly pre-processing early enough. Can be in separate PR :)

Author: I avoided it for now, as spec is unstable, and our implementation is also experimental. But we can add it in the future as we design this.
